Introduce a churnOp to scheduler perf testing framework

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind feature
/sig scheduling

**What this PR does / why we need it**:

Introduce a churn operator to scheduler perf testing framework. The churn operator is aimed to provide configurable options to continuously churn the cluster by creating/recreating API objects. Below is an explanation for new fields:

- **opcode**: the value must be `churn`
- **templatePaths**: an array specifying spec files for different API objects. For instance:
    ```yaml
    templatePaths:
    - config/churn/node-default.yaml
    - config/churn/service-default.yaml
    ```
- **mode**: two modes are supported for now:
    - **recreate**: creates N objects and then deletes afore-create N objects. N is specified by `number`
    - **create**: continuously create objects until a threshold is met. The threshold is specified by `number`
- **number**: cycles of operations to run-to-stop (for mode `create`) or an infinite create-N-objs-delete-N-objs iteration (for mode `recreate`).

**Which issue(s) this PR fixes**:

Fix the 3rd issue of #98898.

**Special notes for your reviewer**:

Some examples:

- templatePaths = [churn/node.yaml, churn/service.yaml], mode = **recreate**, number = **1**
    ```
    [DEBUG] create name = node-churn-rgx94
    [DEBUG] create name = service-churn-vjcp7
    [DEBUG] create name = node-churn-ztvgz
    [DEBUG] create name = service-churn-c4tnc
    [DEBUG] delete name = node-churn-rgx94
    [DEBUG] delete name = service-churn-vjcp7
    [DEBUG] delete name = node-churn-ztvgz
    [DEBUG] delete name = service-churn-c4tnc
    [DEBUG] create name = node-churn-82lp8
    [DEBUG] create name = service-churn-nmz8l
    [DEBUG] create name = node-churn-lggct
    [DEBUG] create name = service-churn-r6cfr
    [DEBUG] delete name = node-churn-82lp8
    [DEBUG] delete name = service-churn-nmz8l
    [DEBUG] delete name = node-churn-lggct
    [DEBUG] delete name = service-churn-r6cfr
    [DEBUG] create name = node-churn-rzdzv
    [DEBUG] create name = service-churn-g7dgf
    # benchmark test stops.
- templatePaths = [churn/node.yaml, churn/service.yaml], mode = **recreate**, number = **3**
    ```
    [DEBUG] create name = node-churn-52ptg
    [DEBUG] create name = service-churn-nxhqj
    [DEBUG] create name = node-churn-tmvwz
    [DEBUG] create name = service-churn-hxnnz
    [DEBUG] create name = node-churn-lj9w5
    [DEBUG] create name = service-churn-kc5cd
    [DEBUG] delete name = node-churn-52ptg
    [DEBUG] delete name = service-churn-nxhqj
    [DEBUG] delete name = node-churn-tmvwz
    [DEBUG] delete name = service-churn-hxnnz
    [DEBUG] delete name = node-churn-lj9w5
    [DEBUG] delete name = service-churn-kc5cd
    # benchmark test stops.
    ```
- templatePaths = [churn/node.yaml, churn/service.yaml], mode = **create**, number = **0** (0 = infinite)
    ```
    [DEBUG] create name = node-churn-vkvlt
    [DEBUG] create name = service-churn-tbkvf
    [DEBUG] create name = node-churn-hpfp8
    [DEBUG] create name = service-churn-c5xl5
    [DEBUG] create name = node-churn-lddzf
    [DEBUG] create name = service-churn-czxfw
    [DEBUG] create name = node-churn-pnf6x
    [DEBUG] create name = service-churn-fpd97
    [DEBUG] create name = node-churn-mjp4w
    [DEBUG] create name = service-churn-gpssn
    # benchmark test stops.
    ```
- templatePaths = [churn/node.yaml, churn/service.yaml], mode = **create**, number = **2**
    ```
    [DEBUG] create name = node-churn-txcbv
    [DEBUG] create name = service-churn-wlpwj
    [DEBUG] create name = node-churn-fq4xd
    [DEBUG] create name = service-churn-vfv2s
    # benchmark test not stop yet
    # no ouptut from churn opeartor
    # benchmark test stops.
    ```

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

Huang-Wei committed:
Introduce a churnOp to scheduler perf testing framework

- support two modes: recreate and create
- use DynmaicClient to create API objects

Huang-Wei commented:
/assign @adtac

alculquicondor commented:
Why does creating a service generate requeuing? Is this because of a non-default plugin?

If so, this requeuing event won't be very helpful in the future. Rather, we can test the most critical paths: node creation, tolerations change, pod affinity, etc.

Huang-Wei commented:
> Why does creating a service generate requeuing? Is this because of a non-default plugin?

Our current logic moves pods anyways upon service events, so a Service generator can be an easy way to simulate churn.

> If so, this requeuing event won't be very helpful in the future.

Probably, esp. if we remove the service event handlers when ServiceAffinity gets deprecated.

> Rather, we can test the most critical paths: node creation, tolerations change, pod affinity, etc.

We definitely should exercise to introduce those events as well as Volume events, but again, a Service generator can be a good start.

alculquicondor commented:
> Probably, esp. if we remove the service event handlers when ServiceAffinity gets deprecated

Not just that, which is happening soon regardless. With your change, since the default plugins don't include ServiceAffinity, we would effective remove any reactions to service events.

So, what I'm saying is that this test is not very future proof.

Huang-Wei commented:
@alculquicondor I updated the PR to support add/deleting API objects interleavingly. API objects include Nodes/Pods/Services for now.

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -43,6 +44,7 @@ const (
 	configFile        = "config/performance-config.yaml"
 	createNodesOpcode = "createNodes"
 	createPodsOpcode  = "createPods"
+	churnOpcode       = "churn"
```
let's name this in a way that test configurations are more readable. Perhaps `continuosRecreates` or `noiseRecreates`.

Huang-Wei replied:
We used `churn` a lot and it's the most readable wording to me :) But I'm not a naming expert, @ahg-g @adtac any naming thoughts?

adtac replied:
I prefer churn too. Maybe "backgroundChurn" to emphasise that this is non-blocking? would also somewhat cover @alculquicondor's suggestion

alculquicondor replied:
```
  - opcode: churn
    serviceTemplatePath: config/churn/service-default.yaml
    intervalMilliseconds: 1000
```

this doesn't make sense to me. It almost feel like networking churn. Maybe `recreatesChurn`

Huang-Wei replied:
The pure-service churn is an example of evaluating how Service events may or may not churn the cluster. I removed it to postpone the debate of how practical it could be - I'd like to get to a consensus of how a churn op looks like first, and in follow-up PRs we can add/discuss test cases on demand, also how wild the test case can churn the cluster.

The current logic is to create & delete _customizable_ API objects interleavingly. Other options are:

- create & delete **N** API objects interleavingly
- keep creating API objects (without deleting them)

@alculquicondor @adtac WDYT?

alculquicondor replied:
can we have something like this:

```
  - opcode: apiChurn
    mode: recreate
    serviceTemplatePath: config/churn/service-default.yaml
    intervalMilliseconds: 1000
```

Note that I'm not referring to the specific test case, just how the test description and configuration looks like.

Huang-Wei replied:
SG. So we have different churn modes/patterns to choose from.

- recreate (or continuous-create)
- ephemeral-create (create & delete a single object)
- round-robin-create (create N objects, and then delete N objects, N can be configurable)
- and probably incorporate update events

BTW: when you say `recreate`, do you mean continuously creating API objects?

alculquicondor commented on test/integration/scheduler_perf/config/performance-config.yaml:
```diff
@@ -449,3 +449,32 @@
       initPods: 2000
       measurePods: 5000
+
+- name: PendingPodsWithMixedChurn
```
I have another idea, which perhaps is more realistic and something we would like tested. 

Create pods first (they are unschedulable). Then create Nodes (perhaps have a ticker for it). We would have to combine that with pending pods that are unschedulable for different reasons (like waiting for pod affinity).

@adtac didn't we have a test similar to what I'm describing?

alculquicondor replied:
Another very important one:
New nodes show up, but with not-ready taint, then the taint is removed.

Huang-Wei replied:
> Create pods first (they are unschedulable). Then create Nodes (perhaps have a ticker for it).

SG. However, in that case we need to enable both `SkipWaitToCompletion` and `collectMetrics`, which is not supported yet:

https://github.com/kubernetes/kubernetes/blob/b5808c6df97a786047c04a35371b20104a1dd28d/test/integration/scheduler_perf/scheduler_perf_test.go#L379-L381

adtac replied:
>didn't we have a test similar to what I'm describing?

no, I don't think we did. multiple createNodes in one workload was one of my motivations for the scheduler_perf rewrite, but I never got around to opening a PR with such a workloda even though I tested it locally :/

>However, in that case we need to enable both SkipWaitToCompletion and collectMetrics

True. Unfortunately, adding support for specifying both parameters seems to be quite difficult because of how metrics/throughput collection works.

The node taint one is a good idea :+1:  but, of course, it requires a modifyNodes op or something like that.

Huang-Wei commented:
/retest

adtac commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -387,6 +422,98 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+			}
+
+			if len(churnFns) == 0 {
+				b.Fatalf("op %d: at least one churn function needs to be defined", opIndex)
```
we should do this in isValid (check if at least one of {Node,Pod,Service}TemplatePath is nil). doing this in isValid will let the framework exit early instead of at workloda runtime

adtac commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -387,6 +422,98 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+			if concreteOp.PodTemplatePath != nil {
+				pod := &v1.Pod{}
+				if err := getSpecFromFile(concreteOp.PodTemplatePath, pod); err != nil {
+					b.Fatalf("op %d: unable to parse Node: %v", opIndex, err)
```
```suggestion
					b.Fatalf("op %d: unable to parse Pod: %v", opIndex, err)
```

adtac commented on test/integration/scheduler_perf/config/performance-config.yaml:
```diff
@@ -449,3 +449,32 @@
+    countParam: $initPods
+    podTemplatePath: config/pod-high-priority-large-cpu.yaml
+    skipWaitToCompletion: true
+  - opcode: churn
```
should the churn pods be in the same namespace as the createPods op? I can't remember if that makes a difference in the scheduler

Huang-Wei replied:
I assume the users can provide namespace name in the configuration if they think namespace is correlated.

Huang-Wei commented:
Hmm... It appears the pattern of "create-delete" nodes/pods/service obj cannot go through a 5000-nodes test. In particular, it's the "deletion" operation that looks special - I tried to switch the churn operation to continuously create API objects, which works great. In terms of scheduler logic, I cannot find any clues that may relate to "deletion". @adtac @alculquicondor Any thoughts?

BTW: you can give a shot by running:

```
make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/PendingPodsWithServiceChurn/5000Nodes/2000InitPods/1000PodsToSchedule -data-items-dir /tmp"
```

Huang-Wei commented:
/retest

Huang-Wei commented:
Follow-up of https://github.com/kubernetes/kubernetes/pull/98900#discussion_r577934655, I rewrote this PR to provide different modes (currently `create` and `recreate`) to create/recreate API objects, also introduced dynamic client to accept arbitrary spec file ins `templatePaths`.

@alculquicondor @adtac Please see udpated description section for more details.

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -43,7 +50,11 @@ const (
+	churnOpcode       = "churn"
 	barrierOpcode     = "barrier"
+
+	Recreate = "recreate"
```
these ones deserve a comment

Huang-Wei replied:
done.

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -387,6 +448,102 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+					b.Fatalf("op %d: unable to parse the %v-th template path: %v", opIndex, i, err)
+				}
+				// Obtain GVR.
+				mapping, err := restMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
```
Interesting. I didn't know this was possible.

alculquicondor commented on test/integration/scheduler_perf/util.go:
```diff
@@ -55,28 +58,35 @@ var dataItemsDir = flag.String("data-items-dir", "", "destination directory for
+	client := clientset.NewForConfigOrDie(cfg)
+	dynClient := dynamic.NewForConfigOrDie(cfg)
+	cachedClient := cacheddiscovery.NewMemCacheClient(client.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(cachedClient)
```
super nit: calculate `restMapper` outside of this function, as only one test cares about it.

Huang-Wei replied:
done.

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -43,7 +52,14 @@ const (
+	churnOpcode       = "churn"
 	barrierOpcode     = "barrier"
+
+	// Two modes supported in "churn" operator.
```
nit: add empty line, as this comment refers to more than one constant

Huang-Wei replied:
done.

alculquicondor reviewed (commented):
good for squash to me.
Anything to add @adtac?

Huang-Wei commented:
/retest

Huang-Wei commented:
/priority important-soon
/triage accepted

adtac reviewed (commented):
/lgtm
/hold for nit since this is approved, feel free to remove and merge
sorry I couldn't review this earlier, I was busy with the code freeze :) we're still within the test freeze, so this PR should be fine for 1.21

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest
